bser/encode.go

```diff
@@ -190,6 +194,75 @@ func encode(buf []byte, d interface{}) ([]byte, error) {
 	}
 }
 
+func canTemplateEncode(r reflect.Value) bool {
+	if r.Kind() != reflect.Slice && r.Kind() != reflect.Array {
+		return false
+	}
+	elem := r.Type().Elem()
+	if elem.Kind() == reflect.Struct {
+		// slice/array of structs
+		return true
+	}
+
+	if elem.Kind() == reflect.Ptr && elem.Elem().Kind() == reflect.Struct {
+		// slice/array of pointers to structs
+		for i := 0; i < r.Len(); i++ {
+			if r.Index(i).IsNil() {
+				// can't use template encoding if any values nil
+				return false
+			}
+		}
+		return true
+	}
+	return false
+}
+
+func encodeTemplate(buf []byte, r reflect.Value) ([]byte, error) {
+	var (
+		elem  = r.Type().Elem()
+		isPtr = false
+	)
+
+	if elem.Kind() == reflect.Ptr {
+		elem = elem.Elem()
+		isPtr = true
+	}
```

Comment:
I'd like to generalize this a bit so we can use it everywhere we might handle a pointer (not just in slice handling).  Also this should probably handle interfaces as well...

Reply:
@jonasi  why should this handle interfaces? With an interface we have no guarantees that each element will have the same fields

Reply:
that's a good point.  Still would like to generalize ptr handling, but perhaps we can do later.